src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs

```diff
@@ -23,65 +25,242 @@ namespace System.Text.RegularExpressions.Generator
+                    Dictionary<string, string[]> requiredHelpers = new();
+                    var sw = new StringWriter();
+                    var writer = new IndentedTextWriter(sw);
+                    writer.Indent += 3;
+                    writer.WriteLine();
+                    EmitRegexDerivedTypeRunnerFactory(writer, regexType.Method, requiredHelpers);
+                    writer.Indent -= 3;
+                    return (regexType, sw.ToString(), requiredHelpers);
                 })
                 .Collect();
 
+            // To avoid invalidating every regex's output when anything from the compilation changes,
+            // we extract from it the only things we care about: whether unsafe code is allowed,
+            // and a name based on the assembly's name, and only that information is then fed into
+            // RegisterSourceOutput along with all of the cached generated data from each regex.
+            IncrementalValueProvider<(bool AllowUnsafe, string? AssemblyName)> compilationDataProvider =
+                context.CompilationProvider
+                .Select((x, _) => (x.Options is CSharpCompilationOptions { AllowUnsafe: true }, x.AssemblyName));
+
             // When there something to output, take all the generated strings and concatenate them to output,
             // and raise all of the created diagnostics.
-            context.RegisterSourceOutput(codeOrDiagnostics, static (context, results) =>
+            context.RegisterSourceOutput(codeOrDiagnostics.Combine(compilationDataProvider), static (context, compilationDataAndResults) =>
             {
-                var code = new List<string>(s_headers.Length + results.Length);
+                ImmutableArray<object?> results = compilationDataAndResults.Left;
+
+                // Report any top-level diagnostics.
+                bool allFailures = true;
+                foreach (object? result in results)
+                {
+                    if (result is Diagnostic d)
+                    {
+                        context.ReportDiagnostic(d);
+                    }
+                    else
+                    {
+                        allFailures = false;
+                    }
+                }
+                if (allFailures)
+                {
+                    return;
+                }
+
+                // At this point we'll be emitting code.  Create a writer to hold it all.
+                var sw = new StringWriter();
+                IndentedTextWriter writer = new(sw);
+
+                // Add file headers and required usings.
+                foreach (string header in s_headers)
+                {
+                    writer.WriteLine(header);
+                }
+                writer.WriteLine();
+
+                // For every generated type, we give it an incrementally increasing ID, in order to create
+                // unique type names even in situations where method names were the same, while also keeping
```

Comment:
I must be missing something obvious here, but how can two regex generated methods be the same? If one has parameters then I believe the generator will just fail to emit anything, so two methods could only be named the same if they are on a different type, which wouldn't be a problem in our case, so I guess I'm not understanding why the id is necessary

Reply:
> so two methods could only be named the same if they are on a different type, which wouldn't be a problem in our case

Why wouldn't it be a problem?

Reply:
I was just thinking about it in the wrong way, I was basically thinking about:
```c#
partial class A
{
    public static partial global::System.Text.RegularExpressions.Regex MyRegex() => GeneratedClassForA.Instance;
}

partial class B
{
    public static partial global::System.Text.RegularExpressions.Regex MyRegex() => GeneratedClassForB.Instance;
}
```

But I now understand that the Generated class that holds the types with the instances is the same for all regexes, and the name of those inner types are just the method name without includding original containing type name (in my example, `A` or `B`)

Reply:
In case we don't have one already, it might make sense to add a test that has two types with the same Regex method name to ensure this won't get broken in the future, specially since this scenario wouldn't be a problem before this change.

Reply:
> it might make sense to add a test

I don't think we have one.  I'll add one.